text/0860-ember-data-request-service.md

```diff
@@ -0,0 +1,585 @@
+resolve until the response stream is fully read.
+
+```ts
+interface Future<T> extends Promise<StructuredDocument<T>> {
+  abort(): void;
+
+  async getStream(): ReadableStream | null;
+}
+```
+
+The `StructuredDocument` interface is the same as is proposed in RFC 854 and copied below:
+
+```ts
+interface StructuredDocument<T> {
+  request: {
+    url: string;
+    cache?: { key?: string, reload?: boolean, backgroundReload?: boolean };
+    method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
+    data?: Record<string, unknown>;
+    options?: Record<string, unknown>;
+    headers: Record<string, string>;
+  }
+  response: {
+    status: HTTPStatusCode;
+    headers: Record<string, string>;
+  }
+  data: T;
+  error?: Error;
+}
+```
+
+**Request Handlers**
+
+Requests are fulfilled by handlers. A handler receives the request context
+as well as a `next` function with which to pass along a request to the next
+handler if it so chooses.
+
+If a handler calls `next`, it receives a `Future` which resolves to a `StructuredDocument`
+that it can then compose how it sees fit with its own response.
+
+```ts
+
+type NextFn<P> = (req: RequestInfo) => Future<P>;
+
+interface Handler<T> {
+  async request(context: RequestContext, next: NextFn<P>): T;
+}
+```
+
+`RequestContext` contains information about the request as well as a few methods
+ for building up the `StructuredDocument` and `Future` that will be part of the
+ response.
+
+```ts
+interface RequestContext<T> {
+  readonly request: RequestInfo;
+
+  setStream(stream: ReadableStream | Promise<ReadableStream>): void;
+  setResponse(response: Response): void;
+}
+
+interface RequestInfo {
+  url: string;
+  method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
+  data?: Record<string, unknown>;
+  options?: Record<string, unknown>;
+  headers: Record<string, string>;
+  signal: AbortSignal;
+}
+```
+
+A basic `fetch` handler with support for streaming content updates while
+the download is still underway might look like the following, where we use
+[`response.clone()`](https://developer.mozilla.org/en-US/docs/Web/API/Response/clone) to `tee` the `ReadableStream` into two streams.
+
+A more efficient handler might read from the response stream, building up the 
+response data before passing along the chunk downstream.
+
+```ts
+const FetchHandler = {
+  async request(context) {
+    const response = await fetch(context.request);
+    context.setResponse(reponse);
+    context.setStream(response.clone().body);
+
+    return response.json();
+  }
+}
+```
+
+Request handlers are registered by configuring the manager via `use`
+
+```ts
+manager.use([Handler1, Handler2])
+```
+
+Handlers will be invoked in the order they are registered ("fifo", first-in 
+first-out), and may only be registered up until the first request is made. It
+is recommended but not required to register all handlers at one time in order
+to ensure explicitly visible handler ordering.
```

Comment:
In some frameworks with a list of handlers invoked in a particular order (e.g. Ruby Rack), there are often methods such as:

- insert
- insert_before
- move_after

They all help adding in additional "functions" at a specific index.

Could be useful here too. But something that can easily be saved for a later revision (not needed for v1). However, maybe `insert` would be a better name than `use`, because it's easier to add the sibling methods mentioned above (works better name-wise).

Reply:
Initially a DAG was considered (similar to how initializers and addons can be ordered); however, the trouble with these approaches is while the ordering seems to make sense when viewing the late-add in isolation, the lack of explicit upfront ordering still generates high levels of confusion. If it turns out there's a strong use-case for just-in-time additions of handlers we can consider that in a new RFC.

Reply:
Makes sense! At least for us, I don't see this as a problem. 

Also, makes sense to keep the scope of initial RFC down a bit. As you mention, if this turns out to be a problem for a lot of people, one can always follow up later with additional API surface.